modules/javafx.base/src/main/java/com/sun/javafx/binding/OldValueCachingListenerList.java

```diff
@@ -0,0 +1,181 @@
+     * @param value a value to store, can be {@code null}
+     */
+    public final void putLatestValue(T value) {
+        this.latestValue = value;
+    }
+
+    /**
+     * Notifies all listeners using the given observable value as source.
+     *
+     * @param observableValue an {@link ObservableValue}, cannot be {@code null}
+     * @return {@code true} if the listener list is not locked, and it was modified during
+     *     notification otherwise {@code false}
+     */
+    public boolean notifyListeners(ObservableValue<T> observableValue) {
+        boolean wasLocked = isLocked();
+
+        if (!wasLocked) {
+            lock();
+        }
+
+        int initialProgress = progress;  // save as it will be modified soon
+        int invalidationListenersSize = invalidationListenersSize();
+        int maxInvalidations = wasLocked ? Math.min(initialProgress + 1, invalidationListenersSize) : invalidationListenersSize;
+
+        for (int i = 0; i < maxInvalidations; i++) {
+            InvalidationListener listener = getInvalidationListener(i);
+
+            if (listener == null) {
+                continue;
+            }
+
+            // communicate to a lower level loop (if triggered) how many listeners were notified so far:
+            progress = i;
+
+            // call invalidation listener (and perhaps a nested notification):
+            callInvalidationListener(observableValue, listener);
+        }
+
+        int changeListenersSize = changeListenersSize();
+        int maxChanges = wasLocked ? Math.min(initialProgress + 1 - invalidationListenersSize, changeListenersSize) : changeListenersSize;
+
+        T oldValue = getLatestValue();
+        T newValue = null;
+
+        for (int i = 0; i < maxChanges; i++) {
+            ChangeListener<T> listener = getChangeListener(i);
+
+            if (listener == null) {
+                continue;
+            }
+
+            // only get the latest value if this is the first loop or a nested notification occurred:
+            if (progress < 0 || i == 0) {
+                newValue = observableValue.getValue();
+
+                // Latest value should even be stored if it was "equals", as it may be a different reference
+                putLatestValue(newValue);
+
+                if (Objects.equals(newValue, oldValue)) {
+                    break;
+                }
+            }
+
+            // communicate to a lower level loop (if triggered) how many listeners were notified so far:
+            progress = i + invalidationListenersSize;
+
+            // call change listener (and perhaps a nested notification):
+            callChangeListener(observableValue, listener, oldValue, newValue);
+        }
+
+        // communicate to a higher level loop that a nested notification occurred:
+        progress = -1;
+
+        return wasLocked ? false : unlock();
+    }
+
+    private void callInvalidationListener(ObservableValue<?> instance, InvalidationListener listener) {
```

Comment:
Yes, it would be good to just put these static somewhere. They don't really fit well in the manager class (as the List class which doesn't depend on manager would need to call them there), and the other way around is also odd, given that I only need to call them when not in "list" mode. Still, putting them package private in the ListenerListBase could work....